File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/735-AsteroidCollision/735-AsteroidCollision.cpp
// Last updated: 8/4/2026, 7:13:19 PM
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        int n = asteroids.size();
        vector<int>res;
        stack<int>st;
        

        for(int i = 0 ; i < n ; i++){
            bool destroyed = false;
            while(!st.empty() && st.top() > 0 && asteroids[i] < 0){
                if((abs(st.top())) < (abs(asteroids[i]))){
                    st.pop();
                }
                else if((abs(st.top())) == (abs(asteroids[i]))){
                    st.pop();
                    destroyed = true;
                    break; //both will explode
                }
                else{
                    destroyed = true;
                    break; //asteroids[i] will explode, so break and move to next
                }

            }
            if(!destroyed){
                st.push(asteroids[i]);
            }
        }
        while(!st.empty()){
            res.push_back(st.top());
            st.pop();
        }

        reverse(res.begin(), res.end());
        return res;

    }
};

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/735-AsteroidCollision/735-AsteroidCollision.java
// Last updated: 8/4/2026, 7:13:27 PM
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int n = asteroids.length;
        Stack<Integer>st = new Stack<>();
        
        for(int i = 0 ; i < n ; i++){
            if(asteroids[i] > 0){
                st.push(asteroids[i]);
            }
            else{
                while(!st.empty() && st.peek() > 0 && asteroids[i] < 0){
                    if(st.peek() < -asteroids[i]){
                        st.pop();
                        continue;
                    }
                    else if(st.peek() == -asteroids[i]){
                        st.pop();
                    }
                    asteroids[i] = 0;
                }
                if(asteroids[i] != 0){
                    st.push(asteroids[i]);
                }
            }   
        }
        int[] result = new int[st.size()];
        for(int i = st.size() - 1; i >= 0; i--) {
            result[i] = st.pop();
        }
        return result;

    }
}